runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/TerminatingExecutors.java

```diff
@@ -0,0 +1,206 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.util;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.apache.beam.runners.dataflow.worker.WorkerUncaughtExceptionHandler;
+import org.apache.beam.runners.dataflow.worker.util.common.worker.JvmRuntime;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+
+/**
+ * Utility class for {@link java.util.concurrent.ExecutorService}s that will terminate the JVM on
+ * uncaught exceptions.
+ *
+ * @implNote Ensures that all threads produced by the {@link ExecutorService}s have a {@link
+ *     WorkerUncaughtExceptionHandler} attached to prevent hidden/silent exceptions and errors.
+ */
+public final class TerminatingExecutors {
+  private TerminatingExecutors() {}
+
+  public static TerminatingExecutorService newSingleThreadExecutor(
+      ThreadFactoryBuilder threadFactoryBuilder, Logger logger) {
+    return new TerminatingExecutorService(
+        Executors.newSingleThreadExecutor(terminatingThreadFactory(threadFactoryBuilder, logger)));
+  }
+
+  public static TerminatingScheduledExecutorService newSingleThreadScheduledExecutor(
+      ThreadFactoryBuilder threadFactoryBuilder, Logger logger) {
+    return new TerminatingScheduledExecutorService(
+        Executors.newSingleThreadScheduledExecutor(
+            terminatingThreadFactory(threadFactoryBuilder, logger)));
+  }
+
+  public static TerminatingExecutorService newCachedThreadPool(
+      ThreadFactoryBuilder threadFactoryBuilder, Logger logger) {
+    return new TerminatingExecutorService(
+        Executors.newCachedThreadPool(terminatingThreadFactory(threadFactoryBuilder, logger)));
+  }
+
+  public static TerminatingExecutorService newFixedThreadPool(
+      int numThreads, ThreadFactoryBuilder threadFactoryBuilder, Logger logger) {
+    return new TerminatingExecutorService(
+        Executors.newFixedThreadPool(
+            numThreads, terminatingThreadFactory(threadFactoryBuilder, logger)));
+  }
+
+  public static TerminatingExecutorService newSingleThreadedExecutorForTesting(
+      JvmRuntime jvmRuntime, ThreadFactoryBuilder threadFactoryBuilder, Logger logger) {
+    return new TerminatingExecutorService(
+        Executors.newSingleThreadExecutor(
```

Comment:
I think the test is passing since you have `newSingleThreadExecutor` here. Try changing this to `newSingleThreadScheduledExecutor`, the test will fail.

something like WrappingExecutorService suggested by @scwhittle seems better. We can use the new executor only where ScheduledExecutorService are needed (eg: refresh), other places we can use plain `Executors.newSingleThreadExecutor`

Reply:
sgtm will update

Reply:
do we want to crash the JVM on failures of tasks submitted to ScheduledExecutorService?

Reply:
I think it depends on the use case. There can be cases where the futures are being examined and having the exceptions manually handled by retrying etc.  If we are using schedule and not ignoring the future result then that seems like a case we wouldn't want a failure.  If we are using execute or otherwise ignoring the future it seems like we would want a failure.

I think Arun's advice to use ExecutorService where we can makes sense.  We can look at the remaining scheduledexecutorservice cases to see if we want to fail or if the errors are already handled via examining the futures.

Reply:
updated thanks!